src/processor/operator/persistent/node_batch_insert.cpp

```diff
@@ -135,56 +151,96 @@ void NodeBatchInsert::writeAndResetNodeGroup(node_group_idx_t nodeGroupIdx,
     nodeGroup->resetToEmpty();
 }
 
-void NodeBatchInsertSharedState::calculateNumTuples() {
-    numRows.store(StorageUtils::getStartOffsetOfNodeGroup(getCurNodeGroupIdx()));
-    if (sharedNodeGroup) {
-        numRows += sharedNodeGroup->getNumRows();
+common::offset_t NodeBatchInsert::writeToExistingNodeGroup(transaction::Transaction* transaction,
+    node_group_idx_t nodeGroupIdx, std::optional<IndexBuilder>& indexBuilder,
+    column_id_t pkColumnID, NodeTable* table, ChunkedNodeGroup* nodeGroup) {
+    auto numExistingTuplesInChunk = table->getNumTuplesInNodeGroup(transaction, nodeGroupIdx);
+    auto valuesToWrite = std::min(StorageConstants::NODE_GROUP_SIZE - numExistingTuplesInChunk,
+        nodeGroup->getNumRows());
+    auto nodeOffset =
+        StorageUtils::getStartOffsetOfNodeGroup(nodeGroupIdx) + numExistingTuplesInChunk;
+    if (indexBuilder) {
+        indexBuilder->insert(nodeGroup->getColumnChunkUnsafe(pkColumnID), nodeOffset,
+            valuesToWrite);
     }
+    auto nodeInfo = ku_dynamic_cast<BatchInsertInfo*, NodeBatchInsertInfo*>(info.get());
+    LocalNodeNG localNodeGroup(StorageUtils::getStartOffsetOfNodeGroup(nodeGroupIdx),
+        nodeInfo->columnTypes);
+    auto& insertChunks = localNodeGroup.getInsertChunks();
+    insertChunks.append(numExistingTuplesInChunk, nodeGroup, valuesToWrite);
+    table->prepareCommitNodeGroup(nodeGroupIdx, transaction, &localNodeGroup);
+    return valuesToWrite;
+}
+
+void NodeBatchInsert::clearToIndex(std::unique_ptr<ChunkedNodeGroup>& nodeGroup,
+    common::offset_t startIndexInGroup) {
+    // Create a new chunked node group and move the unwritten values to it
+    // TODO(bmwinger): Can probably re-use the chunk and shift the values
```

Comment:
Maybe. Do you mean write the last `n` values to the existing chunk, truncate the group, and continue adding to the existing one and use that when creating a new node group?
In addition to it meaning that even in single-threaded mode the values will not be in their original order (though I'm not sure if we have any guarantees of that anyway), truncation could lead to things like the re-used `ListColumnChunk` storing extra values from the original data, since `ListColumnChunk::finalize` doesn't always remove unused values.
But either way it would add a function that needs to be implemented by all the column chunk types and could have subtle side-effects if omitted (but not failures, which makes it easier to miss), which was why I decided against including that in this PR.